### ISSUE TITLE
 Cannot find module 'q'

### ISSUE THREAD

**Issue by romanr** [lambduh/lambduh#1]
What am I doing wrong? How to get it working? I tried different functions and examples, all of them fail with same error:

```
{
  "errorMessage": "Cannot find module 'q'",
  "errorType": "Error",
  "stackTrace": [
    "Function.Module._resolveFilename (module.js:338:15)",
    "Function.Module._load (module.js:280:25)",
    "Module.require (module.js:364:17)",
    "require (module.js:380:17)",
    "Object.<anonymous> (/var/task/index.js:1:71)",
    "Module._compile (module.js:456:26)",
    "Object.Module._extensions..js (module.js:474:10)",
    "Module.load (module.js:356:32)",
    "Function.Module._load (module.js:312:12)",
    "Module.require (module.js:364:17)"
  ]
}
```

**Comment by russmatney** [comment]
Assuming you're using something like the example in the readme:

```
var Q = require('q');
```

The above snippet will throw that error if you haven't installed the `q` module. You can do so via `npm install q`. I recommend getting things running on your machine before trying to run it on amazon, you'll save yourself a ton of debugging time.

Let me know if this helps - i'm happy to create an example or two in this repo if it helps you get started.

**Comment by romanr** [comment]
I get this error when I try to run on Aws .I just paste the function in
there. Is there anything else I need to do?
Add "npm install q" in front of the script? That won't work.

**Comment by russmatney** [comment]
You are correct – copying and pasting the example function into Lambda would never work.

A simple function that doesn't require external modules would run fine on there – something like:

``` javascript
exports.handler = function(event, context) {
  console.log(event);
}
```

In node, `require` is a specialized function that will look for a module with a matching name. There are built in modules, like `path` and `fs`, but if `require` can't find one of those, it will look into your project's `node_modules` folder (in this case, for a folder named `q`), [as explained here in node's docs](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders).

Rather than copying and pasting the function as it is, you need to create a fully functioning node project, zip that, and upload the whole package to lambda.

To get started in that direction, and assuming you've installed node and npm, you can create a new project via with the terminal with something like:

``` bash
mkdir my-new-project
cd my-new-project
npm init
npm install --save q lambduh-validate
touch index.js
```

Then write your function in `index.js`, and test it on your own machine like so:

``` bash
node ./index.js
```

Forgive me if I'm way off here - it's been half a year since I've done any direct work with aws lambda, so if there have been big changes, I don't know about them. I highly recommend reading the full [AWS Lambda Developer Guide](http://docs.aws.amazon.com/lambda/latest/dg/welcome.html), but this sounds more like a misunderstanding of node than lambda, which is why I'd recommend getting things working locally before throwing it over the wall – otherwise your dev-loop is going to be quite painful.
